python/lsst/daf/butler/queries/tree/_base.py

```diff
@@ -0,0 +1,188 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = (
+    "QueryTreeBase",
+    "ColumnExpressionBase",
+    "DatasetFieldName",
+    "InvalidQueryError",
+    "DATASET_FIELD_NAMES",
+)
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, ClassVar, Literal, TypeAlias, TypeVar, cast, get_args
+
+import pydantic
+
+from ...column_spec import ColumnType
+
+if TYPE_CHECKING:
+    from ..visitors import ColumnExpressionVisitor
+    from ._column_literal import ColumnLiteral
+    from ._column_set import ColumnSet
+
+
+# Type annotation for string literals that can be used as dataset fields in
+# the public API.  The 'collection' and 'run' fields are string collection
+# names.  Internal interfaces may define other dataset field strings (e.g.
+# collection primary key values) and hence should use `str` rather than this
+# type.
+DatasetFieldName: TypeAlias = Literal["dataset_id", "ingest_date", "run", "collection", "timespan"]
+
+# Tuple of the strings that can be use as dataset fields in public APIs.
+DATASET_FIELD_NAMES: tuple[DatasetFieldName, ...] = tuple(get_args(DatasetFieldName))
+
+_T = TypeVar("_T")
+_L = TypeVar("_L")
+_A = TypeVar("_A")
+_O = TypeVar("_O")
+
+
+class InvalidQueryError(RuntimeError):
+    """Exception raised when a query is not valid."""
+
+
+class QueryTreeBase(pydantic.BaseModel):
+    """Base class for all non-primitive types in a query tree."""
+
+    model_config = pydantic.ConfigDict(frozen=True, extra="forbid", strict=True)
+
+
+class ColumnExpressionBase(QueryTreeBase, ABC):
+    """Base class for objects that represent non-boolean column expressions in
+    a query tree.
+
+    Notes
+    -----
+    This is a closed hierarchy whose concrete, `~typing.final` derived classes
+    are members of the `ColumnExpression` union.  That union should be used in
+    type annotations rather than the technically-open base class.
+    """
+
+    expression_type: str
+    """String literal corresponding to a concrete expression type."""
+
+    is_literal: ClassVar[bool] = False
+    """Whether this expression wraps a literal Python value."""
+
+    @property
+    @abstractmethod
+    def precedence(self) -> int:
+        """Operator precedence for this operation.
+
+        Lower values bind more tightly, so parentheses are needed when printing
+        an expression where an operand has a higher value than the expression
+        itself.
+        """
+        raise NotImplementedError()
```

Comment:
Done (on DM-43146): we now just add parentheses for any kind of nested expression (but do try to avoid them on literals, direct references to columns, and predicates).